akka-http-core/src/main/scala/akka/http/impl/engine/http2/HttpMessageRendering.scala

```diff
@@ -0,0 +1,188 @@
+    if (entity.contentType != ContentTypes.NoContentType)
+      headerPairs += "content-type" -> entity.contentType.toString
+    entity.contentLengthOption.foreach(headerPairs += "content-length" -> _.toString)
+  }
+
+  private[http2] def renderHeaders(
+    headers:  immutable.Seq[HttpHeader],
+    log:      LoggingAdapter,
+    isServer: Boolean
+  ): Seq[(String, String)] = {
+    val headerPairs = new VectorBuilder[(String, String)]()
+    renderHeaders(headers, headerPairs, None, log, isServer)
+    headerPairs.result()
+  }
+
+  /**
+   * Mutates `headerPairs` adding any valid header from `headersSeq`.
+   * @param peerIdHeader a header providing extra information (e.g. vendor and version) about the
+   *                     peer. For example, a User-Agent on the client or a Server header on the server.
+   */
+  private[http2] def renderHeaders(
+    headersSeq:   immutable.Seq[HttpHeader],
+    headerPairs:  VectorBuilder[(String, String)],
+    peerIdHeader: Option[(String, String)],
+    log:          LoggingAdapter,
+    isServer:     Boolean
+  ): Unit = {
+    def suppressionWarning(h: HttpHeader, msg: String): Unit =
+      log.warning("Explicitly set HTTP header '{}' is ignored, {}", h, msg)
+
+    val it = headersSeq.iterator
+    var peerIdSeen, dateSeen = false
+    def addHeader(h: HttpHeader): Unit = headerPairs += h.lowercaseName -> h.value
+
+    while (it.hasNext) {
+      import akka.http.scaladsl.model.headers._
+      val header = it.next()
+      if ((header.renderInResponses && isServer) || (header.renderInRequests && !isServer)) {
+        header match {
+          case x: Server if isServer =>
+            addHeader(x)
+            peerIdSeen = true
+
+          case x: `User-Agent` if !isServer =>
+            addHeader(x)
+            peerIdSeen = true
+
+          case x: Date =>
+            addHeader(x)
+            dateSeen = true
+
+          case x: CustomHeader =>
+            addHeader(x)
+
+          case x: RawHeader if (x is "content-type") || (x is "content-length") || (x is "transfer-encoding") ||
+            (x is "date") || (x is "server") || (x is "user-agent") || (x is "connection") =>
+            suppressionWarning(x, "illegal RawHeader")
+
+          case x: `Content-Length` =>
+            suppressionWarning(x, "explicit `Content-Length` header is not allowed. Use the appropriate HttpEntity subtype.")
+
+          case x: `Content-Type` =>
+            suppressionWarning(x, "explicit `Content-Type` header is not allowed. Set `HttpResponse.entity.contentType` instead.")
+
+          case x: `Transfer-Encoding` =>
+            suppressionWarning(x, "`Transfer-Encoding` header is not allowed for HTTP/2")
+
+          case x: Connection =>
+            suppressionWarning(x, "`Connection` header is not allowed for HTTP/2")
+
+          case x =>
+            addHeader(x)
+        }
+      }
+    }
+
+    if (!dateSeen) {
```

Comment:
The `Date` header should only be added to responses, how comes it didn't fail any tests?

Reply:
Seems to be copy pasta in Http2ClientSpec where the `Date` header is filtered out which is convinient on the server side but shouldn't be done on the client side.

Reply:
I've added a new unit test for `Date` on requests in `HttpMessageRenderingSpec`.